src/main/java/com/nike/cerberus/cache/DatabaseCache.java

```diff
@@ -0,0 +1,234 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.nike.cerberus.cache;
+
+import com.codahale.metrics.Counter;
+import com.google.common.collect.ImmutableMap;
+import com.google.inject.Injector;
+import com.nike.cerberus.server.config.guice.StaticInjector;
+import com.nike.cerberus.service.MetricsService;
+import com.typesafe.config.Config;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.ibatis.builder.InitializingObject;
+import org.apache.ibatis.cache.Cache;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReadWriteLock;
+
+import static com.github.benmanes.caffeine.cache.Caffeine.newBuilder;
+import static java.util.Optional.ofNullable;
+
+/**
+ * TODO rewrite
+ * A simple Caffeine backed cache that auto expires items after a certain time period,
+ * to help us against bursty traffic that does repeat reads.
+ */
+public class DatabaseCache implements Cache, InitializingObject {
+
+  private final Logger log = LoggerFactory.getLogger(getClass());
+  protected final String id;
+  private MetricsService metricsService;
+
+  protected static final String GLOBAL_DATA_TTL_IN_SECONDS = "cms.mybatis.cache.global.dataTtlInSeconds";
+  protected static final String DATA_TTL_IN_SECONDS_OVERRIDE_PATH_TEMPLATE = "cms.mybatis.cache.%s.dataTtlInSeconds";
+  protected static final String GLOBAL_REPEAT_READ_COUNTER_EXPIRE_IN_SECONDS = "cms.mybatis.cache.global.repeatReadCounterResetInSeconds";
+  protected static final String REPEAT_READ_COUNTER_EXPIRE_IN_SECONDS_OVERRIDE_PATH_TEMPLATE = "cms.mybatis.cache.%s.repeatReadCounterResetInSeconds";
+  protected static final String GLOBAL_REPEAT_READ_THRESHOLD = "cms.mybatis.cache.global.repeatReadThreshold";
+  protected static final String REPEAT_READ_THRESHOLD_OVERRIDE_PATH_TEMPLATE = "cms.mybatis.cache.%s.repeatReadThreshold";
+  protected static final int DEFAULT_GLOBAL_DATA_TTL_IN_SECONDS = 10;
+  protected static final int DEFAULT_REPEAT_READ_COUNTER_EXPIRE_IN_SECONDS = 1;
+  protected static final int DEFAULT_REPEAT_READ_THRESHOLD = 5;
+
+  protected com.github.benmanes.caffeine.cache.Cache<Object, Object> dataCache;
+  protected com.github.benmanes.caffeine.cache.Cache<Object, Counter> autoExpiringRepeatReadCounterMap;
+  protected Integer repeatReadThreshold;
+  protected Counter hitCounter;
+  protected Counter missCounter;
+
+  public DatabaseCache(String id) {
+    this.id = id;
+  }
+
+  /**
+   * This method gets called after this class is instantiated by MyBatis and all the properties have been set.
+   */
+  @Override
+  public void initialize() {
+    // Util we can get the MyBatis Guice Module updated, this is our best bet, for getting the Guice instances.
+    // https://groups.google.com/forum/#!msg/mybatis-user/Ekd1LTNVIDc/t2xGuvETBgAJ
```

Comment:
There is an update on that thread with a better way.

- [ ] TODO, investigate custom config as per the thread.